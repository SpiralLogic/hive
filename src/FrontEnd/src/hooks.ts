import { HiveEvent, HiveEventEmitter, HiveEventListener } from './hive-event-emitter';
import { Inputs, useEffect, useReducer } from 'preact/hooks';

const hiveEventEmitter = new HiveEventEmitter();

export const useHiveEventEmitter = (
  hiveEventHandler?: HiveEventListener<HiveEvent>,
  inputs?: Inputs
): HiveEventEmitter => {
  if (hiveEventHandler)
    useEffect(() => {
      hiveEventEmitter.add(hiveEventHandler);
      return () => {
        hiveEventEmitter.remove(hiveEventHandler);
      };
    }, inputs);

  return hiveEventEmitter;
};

const classReducer = (
  initialClasses: Array<string>,
  action: { type: 'add' | 'remove'; classes: string[] }
): string[] => {
  const { type, classes } = action;
  switch (type) {
    case 'add':
      return [...initialClasses, ...classes.filter((c) => !initialClasses.includes(c))];
    case 'remove':
      return initialClasses.filter((c) => !classes.includes(c));
  }
};

export const useClassReducer = (
  initialClasses: string[]
): [Array<string>, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
