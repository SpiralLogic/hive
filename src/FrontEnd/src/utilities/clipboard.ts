const fallbackCopyTextToClipboard = (text: string) => {
  const currentFocus = document.activeElement as HTMLElement;
  const textArea = document.createElement('textarea');
  textArea.value = text;

  textArea.style.top = '0';
  textArea.style.left = '0';
  textArea.style.position = 'fixed';

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();
  try {
    document.execCommand('copy');
    if (currentFocus && currentFocus.focus) currentFocus.focus();
    return true;
  } catch (e) {
    //  if (process.env.NODE_ENV !== 'production') console.error(e);
  } finally {
    document.body.removeChild(textArea);
  }
  return false;
};

export const shareGame = async (url: string) => {
  const opponentGame = {
    title: 'Hive board game',
    text: 'Join me in a game of the Hive!',
    url,
  };
  try {
    await navigator.share(opponentGame);
    return false;
  } catch {
    if (navigator.clipboard && navigator.clipboard.writeText) {
      try {
        await navigator.clipboard.writeText(opponentGame.url);
        return true;
      } catch {
        return fallbackCopyTextToClipboard(opponentGame.url);
      }
    }
  }
  return false;
};
