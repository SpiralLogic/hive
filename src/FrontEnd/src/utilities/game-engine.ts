import { GameConnection, GameHandlers, HexEngine, OpponentSelectionHandler } from '../domain/engine';
import { GameId, GameState, Move, Tile } from '../domain';
import { HubConnection, HubConnectionBuilder, HubConnectionState } from '@microsoft/signalr';

const requestHeaders = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
};

const moveTile = async (gameId: GameId, move: Move): Promise<GameState> => {
  const response = await fetch(`/api/move/${gameId}`, {
    method: 'POST',
    headers: requestHeaders,
    body: JSON.stringify(move),
  });

  return await response.json();
};

const getExistingGame = async (gameId: GameId): Promise<GameState> => {
  const response = await fetch(`/api/game/${gameId}`, {
    method: `GET`,
    headers: requestHeaders,
  });

  return response.json();
};

const getNewGame = async (): Promise<GameState> => {
  const response = await fetch(`/api/new`, {
    method: `POST`,
    headers: requestHeaders,
    body: JSON.stringify(''),
  });
  return response.json();
};

const connectGame = (gameId: GameId, handlers: GameHandlers): GameConnection => {
  const getConnection = (gameId: GameId): HubConnection => {
    const hubUrl = `${window.location.protocol}//${window.location.host}/gamehub/${gameId}`;
    const connection = new HubConnectionBuilder().withUrl(hubUrl).withAutomaticReconnect().build();
    connection.on('ReceiveGameState', handlers.updateGameState);
    if (handlers.opponentSelection) connection.on('OpponentSelection', handlers.opponentSelection);
    connection.start().then();
    return connection;
  };

  const connection = getConnection(gameId);

  const sendSelection: OpponentSelectionHandler = (type: 'select' | 'deselect', tile: Tile) =>
    connection.state === HubConnectionState.Connected &&
    connection.invoke('SendSelection', type, tile).catch(function (err) {
      return console.error(err.toString());
    });

  const closeConnection = () => {
    connection.off('ReceiveGameState', handlers.updateGameState);
    return connection.stop();
  };

  if (process.env.NODE_ENV !== 'production') {
    connection.onreconnecting((error) => console.warn(`reconnecting to game ${gameId} .. ${error}`));
    connection.onclose((error) => console.info(`connection closed to game ${gameId} .. ${error}`));
    connection.onreconnected((error) => {
      window.location.reload();
      console.info(`reconnected to game ${gameId} .. ${error}`);
    });
  }

  return {
    getConnectionState: () => connection.state,
    sendSelection,
    closeConnection,
  };
};

const Engine: HexEngine = {
  getNewGame,
  getExistingGame,
  moveTile,
  connectGame: connectGame,
};

export default Engine;
