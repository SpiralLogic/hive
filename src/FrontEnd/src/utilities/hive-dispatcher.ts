import { Move, Tile } from '../domain';

export class HiveDispatcher {
  private listeners = new Map<string, Set<HiveEventListener<HiveIntent>>>();

  add = <T extends HiveIntent>(type: T['type'], listener: HiveEventListener<T>): (() => void) => {
    if (!this.listeners.get(type)?.add(listener)) this.listeners.set(type, new Set([listener]));
    return () => this.remove<T>(type, listener);
  };

  dispatch = <T extends HiveIntent>(intent: T): void => {
    this.listeners.get(intent.type)?.forEach((l): void => l(intent));
  };

  remove = <T extends HiveIntent>(type: T['type'], listener: HiveEventListener<T>): void => {
    this.listeners.get(type)?.delete(listener);
  };
}

export type HiveEventListener<T extends HiveIntent> = <HiveIntent extends T>(intent: HiveIntent) => void;
export type HiveIntent = HiveEvent | HiveAction;
export type HiveEvent = MoveEvent | TileEvent;
export type HiveAction = Action | TileAction;
export type MoveEvent = { type: 'move'; move: Move };
export type Action = {
  type: 'tileClear';
};
export type TileEvent = {
  type: 'tileDropped' | 'click' | 'tileSelected' | 'tileDeselected';
  tile: Tile;
};
export type TileAction = {
  type: 'tileSelect' | 'tileDeselect';
  tile: Tile;
};
