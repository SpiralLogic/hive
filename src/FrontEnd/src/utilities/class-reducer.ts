import { useReducer } from 'preact/hooks';

const classReducer = (
  initialClasses: string,
  action: { type: 'add' | 'remove'; classes: string[] }
): string => {
  const classList = new Set(initialClasses.split(' '));
  switch (action.type) {
    case `add`:
      action.classes.forEach((c) => classList.add(c));
      break;
    case 'remove':
      action.classes.forEach((c) => classList.delete(c));
      break;
    default:
      break;
  }
  classList.delete('');
  return Array.from(classList).join(' ');
};

export const useClassReducer = (
  initialClasses: string
): [string, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
