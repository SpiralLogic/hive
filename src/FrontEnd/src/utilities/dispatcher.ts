import { useEffect } from 'preact/hooks';
import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => hiveDispatcher;

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const emitter = useHiveDispatcher();
    return emitter.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;
