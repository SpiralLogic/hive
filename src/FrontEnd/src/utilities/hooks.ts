import { HiveDispatcher, HiveEventListener, HiveIntent } from './hive-dispatcher';
import { useEffect, useReducer } from 'preact/hooks';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => {
  return hiveDispatcher;
};

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const emitter = useHiveDispatcher();
    return emitter.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;

const classReducer = (
  initialClasses: string,
  action: { type: 'add' | 'remove'; classes: string[] }
): string => {
  const classList = new Set(initialClasses.split(' '));
  if (action.type === `add`) {
    action.classes.forEach((c) => classList.add(c));
  } else if (action.type === 'remove') {
    action.classes.forEach((c) => classList.delete(c));
  }
  classList.delete('');
  return Array.from(classList).join(' ');
};

export const useClassReducer = (
  initialClasses: string
): [string, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
