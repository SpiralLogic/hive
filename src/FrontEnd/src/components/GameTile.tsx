import { FunctionComponent, h } from 'preact';
import { Tile as TileType } from '../domain';

import { TileAction } from '../services';
import { addHiveDispatchListener, dispatchHiveEvent, useClassReducer } from '../utilities/hooks';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Tile from './Tile';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex][role="cell"]`;
const playerSelector = `[tabindex].name`;

const GameTile: FunctionComponent<TileType> = (tile: TileType) => {
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex`);

  const deselect = () => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', classes: ['selected'] });
    dispatchHiveEvent({ type: 'tileDeselected', tile: tile });
  };

  const select = () => {
    if (classes.includes('selected')) return;
    dispatchHiveEvent({ type: 'tileClear' });
    setClassList({ type: 'add', classes: ['selected'] });
    dispatchHiveEvent({ type: 'tileSelected', tile: tile });
  };

  addHiveDispatchListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id === id) select();
  });

  addHiveDispatchListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id === id) deselect();
  });

  addHiveDispatchListener('tileClear', () => {
    deselect();
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', classes: ['beforeDrag'] });
    setTimeout(() => setClassList({ type: 'remove', classes: ['beforeDrag'] }), 1);
  };
  const handleDragEnd = () => {
    dispatchHiveEvent({ type: 'tileClear' });
    dispatchHiveEvent({ type: 'tileDropped', tile: tile });
  };
  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    !classes.includes('selected') ? select() : deselect();
  };
  const handleKeyDown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    e.stopPropagation();
    if (!classes.includes('selected')) {
      select();
      setFocus(cellSelector);
    } else {
      deselect();
    }
  };
  const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

  useEffect(() => {
    if (!focus) return;
    const focusElement =
      document.querySelector<HTMLElement>(focus) || document.querySelector<HTMLElement>(playerSelector);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: creature,
    class: classes ,
    draggable: !!moves.length,
    tabindex: moves.length ? 0 : undefined,
  };
  const handlers = moves.length
    ? {
        onclick: handleClick,
        ondragstart: handleDragStart,
        ondragend: handleDragEnd,
        onkeydown: handleKeyDown,
        onmouseleave: handleMouseLeave,
        ondrop: handleDrop,
      }
    : {
        ondrop: handleDrop,
      };
  return <Tile creature={creature} {...attributes} {...handlers} />;
};
GameTile.displayName = 'Tile';
export default GameTile;
