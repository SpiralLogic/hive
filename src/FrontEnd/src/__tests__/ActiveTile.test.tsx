import { HiveAction, HiveEvent, TileEvent } from '../utilities/hive-dispatcher';
import { fireEvent } from '@testing-library/preact';
import { h } from 'preact';
import { renderElement, simulateEvent } from './helpers';
import { useHiveDispatcher } from '../utilities/hooks';
import GameTile from '../components/GameTile';

describe('Tile Tests', () => {
  const tileCanMove = {
    id: 1,
    playerId: 1,
    creature: 'ant',
    moves: [{ q: 1, r: 1 }],
  };
  const tileNoMove = { id: 2, playerId: 0, creature: 'fly', moves: [] };

  const createTileCanMove = () => {
    return renderElement(<GameTile {...tileCanMove} />);
  };

  const createTileNoMove = () => {
    return renderElement(<GameTile {...tileNoMove} />);
  };

  const expectedHiveEvent: HiveAction = {
    type: 'tileSelect',
    tile: tileCanMove,
  };

  describe('Tile render', () => {
    test('has creature', () => {
      expect(createTileNoMove().querySelectorAll('.creature')).toHaveLength(1);
      expect(createTileCanMove().querySelectorAll('.creature')).toHaveLength(1);
    });
  });

  describe('tile events', () => {
    test('click emits tile start event', () => {
      jest.spyOn(useHiveDispatcher(), 'dispatch');
      fireEvent.click(createTileCanMove());
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(useHiveDispatcher().dispatch).toHaveBeenCalledWith(expect.objectContaining(expectedEvent));
    });

    test('enter emits tile start event', () => {
      jest.spyOn(useHiveDispatcher(), 'dispatch');
      fireEvent.keyDown(createTileCanMove(), { key: 'Enter' });
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(useHiveDispatcher().dispatch).toHaveBeenCalledWith(expect.objectContaining(expectedEvent));
    });

    test('enter selects tile', () => {
      const tile = createTileCanMove();
      fireEvent.keyDown(tile, { key: 'Enter' });

      expect(tile).toHaveClass('selected');
    });

    test('second enter deselects tile', () => {
      const tile = createTileCanMove();
      fireEvent.keyDown(tile, { key: 'Enter' });
      fireEvent.keyDown(tile, { key: 'Enter' });

      expect(tile).not.toHaveClass('selected');
    });

    test('arrow keys use handler', () => {
      const tile = createTileCanMove();
      jest.spyOn(tile, 'focus');
      fireEvent.keyDown(tile, { key: 'ArrowDown' });

      expect(tile.focus).toHaveBeenCalled();
    });

    test('Space emits tile start event', () => {
      jest.spyOn(useHiveDispatcher(), 'dispatch');
      fireEvent.keyDown(createTileCanMove(), { key: ' ' });
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(useHiveDispatcher().dispatch).toHaveBeenCalledWith(expect.objectContaining(expectedEvent));
    });

    test('click deselects previous selected tile', () => {
      jest.spyOn(useHiveDispatcher(), 'dispatch');
      const tile = createTileCanMove();
      fireEvent.click(tile);

      expect(tile).toHaveClass('selected');
    });

    test('clicking same tile doesnt fire a tile start event', () => {
      const mock = jest.spyOn(useHiveDispatcher(), 'dispatch');
      const tile = createTileCanMove();
      fireEvent.click(tile);

      mock.mockClear();
      fireEvent.click(tile);

      expect(useHiveDispatcher().dispatch).not.toHaveBeenCalledWith(
        expect.objectContaining({ type: 'tileSelect' })
      );
    });

    test('mouseLeave activates blur', () => {
      const tile = createTileCanMove();
      tile.focus();
      fireEvent.mouseLeave(tile);
      expect(tile).not.toHaveFocus();
    });

    test('is draggable when there are available moves', () => {
      expect(createTileCanMove()).toHaveAttribute('draggable', 'true');
    });

    test('is *not* draggable when there are no moves available', () => {
      expect(createTileNoMove()).toHaveAttribute('draggable', 'false');
    });

    test('on drag emits start event', () => {
      jest.spyOn(useHiveDispatcher(), 'dispatch');
      fireEvent.dragStart(createTileCanMove());
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(useHiveDispatcher().dispatch).toHaveBeenCalledWith(expect.objectContaining(expectedEvent));
    });

    test('on dragEnd emits end event', () => {
      jest.spyOn(useHiveDispatcher(), 'dispatch');
      fireEvent.dragEnd(createTileCanMove());
      const expectedEvent: HiveEvent = {
        type: 'tileDropped',
        tile: tileCanMove,
      };

      expect(useHiveDispatcher().dispatch).toHaveBeenCalledWith(expect.objectContaining(expectedEvent));
    });

    test('default on drop is prevented', () => {
      expect(simulateEvent(createTileCanMove(), 'drop')).toHaveBeenCalled();
      expect(simulateEvent(createTileNoMove(), 'drop')).toHaveBeenCalled();
    });
  });

  describe('Tile Snapshot', () => {
    test('can move matches current snapshot', () => {
      expect(createTileCanMove()).toMatchSnapshot();
    });

    test('no moves matches current snapshot', () => {
      expect(createTileNoMove()).toMatchSnapshot();
    });
  });
});
