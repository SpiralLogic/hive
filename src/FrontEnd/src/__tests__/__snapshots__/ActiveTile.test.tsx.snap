// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile Tests Tile Snapshot can move matches current snapshot 1`] = `
<div
  class="player1 hex tile tile creature"
  draggable="true"
  tabindex="0"
  title="ant"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature ant"
      href="#ant"
    />
  </svg>
</div>
`;

exports[`Tile Tests Tile Snapshot no moves matches current snapshot 1`] = `
<div
  class="player0 hex tile tile creature"
  draggable="false"
  title="fly"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature fly"
      href="#fly"
    />
  </svg>
</div>
`;
